lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/SanitizeAndSplitBam.pm

```diff
@@ -0,0 +1,303 @@
+            $original_flagstat->{total_reads}, $self->_metrics->{input},
+        );
+    }
+
+    if ( $read_count != $metrics->{output}  ) {
+        $self->fatal_message(
+            'Conflicting read counts from output flagstats and output metrics! %s <=> %s',
+            $read_count, $metrics->{output},
+        );
+    }
+
+    $self->debug_message('Verify read count...done');
+    return 1;
+}
+
+sub _write_reads {
+    my $self = shift;
+
+    my ($read1, $read2) =_separate_reads(@_);
+    my @separated_reads = grep { defined } ( $read1, $read2 );
+    return 0 if not @separated_reads;
+
+    my $type = _determine_type_and_set_flags($read1, $read2);
+    my $rg_id = _read_group_id_for_reads(@separated_reads);
+    my $fh = $self->_get_fh_for_read_group_and_type($rg_id, $type);
+    for my $read_tokens ( @separated_reads ) {
+        # Sanitize!
+        _sanitize_read($read_tokens);
+        # Add RG tag
+        push @$read_tokens, 'RG:Z:'.$self->old_and_new_read_group_ids->{$rg_id}->{$type};
+        $fh->print( join( "\t", @$read_tokens)."\n");
+    }
+
+    return @separated_reads;
+}
+
+sub _read_group_id_for_reads {
+    my @rg_ids;
+    for ( @_ ) {
+        my $rg_tag = List::Util::first { $_ =~ m/^RG:/ } @$_;
+        next if not defined $rg_tag;
+        push @rg_ids, (split(':', $rg_tag))[2];
+    }
+
+    return 'unknown' if not @rg_ids;
+    return (List::MoreUtils::uniq(@rg_ids))[0];
+}
+
+sub _separate_reads {
+    # Separate reads into 1/2 removing supplementary but including the secondary alignments
+    my (@read1s, @read2s);
+    for my $read ( @_ ) {
+        # Remove Supplementary 2048 0x800
+        next if $read->[1] & 0x800;
+        if ( $read->[1] & 0x80 ) {
+            # read 2
+            push @read2s, $read; # flagged read2
+        }
+        else {
+            push @read1s, $read; # unflagged reads will be read1
+        }
+    }
+
+    return ( $read1s[0], $read2s[0] );
+}
+
+sub _determine_type_and_set_flags {
+    # Determine the type paired/singleton and set the flags accordingly
+    my $type;
+    # paired will get:
+    # 64/128 first/second in pair
+    # 1      read paired
+    # 4      read unmapped
+    # 8      mate unmapped
+    if ( defined $_[0] and defined $_[1] ) {
+        $type = 'paired';
+        $_[0]->[1] = 77;
+        $_[1]->[1] = 141;
+    }
+    # sigleton will get:
+    # 4  first in pair
+    # 64 read unmapped
+    elsif ( defined $_[0] ) {
+        $type = 'singleton';
+        $_[0]->[1] = 68;
+    }
+    elsif ( defined $_[1] ) {
+        $type = 'singleton';
+        $_[1]->[1] = 68;
+    }
+    # Gotta send in reads!
+    else {
+        die 'No reads given to _determine_type_and_set_flags!';
+    }
+    $type;
+}
+
+sub _sanitize_read {
+    # Upcase sequence
+    $_[0]->[9] = uc $_[0]->[9];
+    # If listed as revcomp, revcomp seq and qual
+    if ( $_[0]->[1] & 0x10 ) {
+        $_[0]->[9] = reverse $_[0]->[9];
+        $_[0]->[9] =~ tr/ATCG/TAGC/;
```

Comment:
It seems like we ought to have a central reverse complement implementation somewhere!  A `grep` of the code finds we have a few different implementations around, though none of them are in a convenient spot.